src/Symfony/Component/ExpressionLanguage/Node/BinaryNode.php

```diff
@@ -52,6 +52,10 @@ public function compile(Compiler $compiler): void
         if ('matches' == $operator) {
             if ($this->nodes['right'] instanceof ConstantNode) {
                 $this->evaluateMatches($this->nodes['right']->evaluate([], []), '');
+            } elseif ($this->nodes['right'] instanceof self) {
+                if ('~' !== $this->nodes['right']->attributes['operator']) {
+                    throw new SyntaxError('A non-string is passed to "matches" instead of a regexp.');
```

Comment:
```suggestion
                    throw new SyntaxError(''The regex passed to "matches" must be a string.'');
```

Reply:
Thanks, sounds much better. fixed in [3faa641](https://github.com/symfony/symfony/pull/58261/commits/3faa641bc15988b6668459074a691777fb07e03e)